vcd/resource_vcd_vm_internal_disk.go

```diff
@@ -0,0 +1,469 @@
+			return fmt.Errorf(errorCompletingTask, err)
+		}
+	}
+	return nil
+}
+
+// resourceVmInternalDiskDelete deletes disk from VM
+func resourceVmInternalDiskDelete(d *schema.ResourceData, m interface{}) error {
+	vcdClient := m.(*VCDClient)
+
+	vcdClient.lockParentVm(d)
+	defer vcdClient.unLockParentVm(d)
+
+	vm, _, err := getVm(vcdClient, d)
+	if err != nil {
+		return err
+	}
+
+	err = powerOffIfNeeded(d, vm)
+	if err != nil {
+		return err
+	}
+
+	err = vm.DeleteInternalDiskById(d.Id())
+	if err != nil {
+		return fmt.Errorf("[Error] failed to delete internal disk: %s", err)
+	}
+
+	err = powerOnIfNeeded(d, vm)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[TRACE] VM internal disk %s deleted", d.Id())
+	d.SetId("")
+	return nil
+}
+
+func getVm(vcdClient *VCDClient, d *schema.ResourceData) (*govcd.VM, *govcd.Vdc, error) {
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return nil, nil, fmt.Errorf(errorRetrievingOrgAndVdc, err)
+	}
+	vapp, err := vdc.GetVAppByName(d.Get("vapp_name").(string), false)
+	if err != nil {
+		return nil, nil, fmt.Errorf("[Error] failed to get vApp: %s", err)
+	}
+	vm, err := vapp.GetVMByName(d.Get("vm_name").(string), false)
+	if err != nil {
+		return nil, nil, fmt.Errorf("[Error] failed to get VM: %s", err)
+	}
+	return vm, vdc, err
+}
+
+// Update the resource
+func resourceVmInternalDiskUpdate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[TRACE] Update Internal Disk with ID: %s started.", d.Id())
+	vcdClient := meta.(*VCDClient)
+
+	vcdClient.lockParentVm(d)
+	defer vcdClient.unLockParentVm(d)
+
+	vm, vdc, err := getVm(vcdClient, d)
+	if err != nil {
+		return err
+	}
+
+	diskSettingsToUpdate, err := vm.GetInternalDiskById(d.Id(), false)
+	if err != nil {
+		return err
+	}
+	log.Printf("[TRACE] Internal Disk with id %s found", d.Id())
+	iops := int64(d.Get("iops").(int))
+	isThinProvisioned := d.Get("thin_provisioned").(bool)
+	diskSettingsToUpdate.Iops = &iops
+	diskSettingsToUpdate.ThinProvisioned = &isThinProvisioned
+	diskSettingsToUpdate.UnitNumber = d.Get("unit_number").(int)
+	diskSettingsToUpdate.BusNumber = d.Get("bus_number").(int)
+	diskSettingsToUpdate.AdapterType = internalDiskBusTypes[d.Get("bus_type").(string)]
+	diskSettingsToUpdate.SizeMb = int64(d.Get("size_in_mb").(int))
+
+	var storageProfilePrt *types.Reference
+	var overrideVmDefault bool
+
+	storageProfileName := d.Get("storage_profile").(string)
+	if storageProfileName != "" {
+		storageProfile, err := vdc.FindStorageProfileReference(storageProfileName)
+		if err != nil {
+			return fmt.Errorf("[Error] error retrieving storage profile %s : %s", storageProfileName, err)
+		}
+		storageProfilePrt = &storageProfile
+		overrideVmDefault = true
+	} else {
+		storageProfilePrt = vm.VM.StorageProfile
+		overrideVmDefault = false
+	}
+
+	diskSettingsToUpdate.StorageProfile = storageProfilePrt
+	diskSettingsToUpdate.OverrideVmDefault = overrideVmDefault
+
+	err = powerOffIfNeeded(d, vm)
```

Comment:
According to my testing and analysis of this code, this call to `powerOffIfNeeded` refreshes `vm` and thus the `DiskSettings` struct referenced by `diskSettingsToUpdate` is no longer part of `vm.VM.VmSpecSection` which is later on used to send the "update" request.
As a result of this the following call to `vm.UpdateInternalDisks` does not update the internal disks at all.

Here is my suggestion, reordering of the method calls:
```suggestion
// Update the resource
func resourceVmInternalDiskUpdate(d *schema.ResourceData, meta interface{}) error {
	log.Printf("[TRACE] Update Internal Disk with ID: %s started.", d.Id())
	vcdClient := meta.(*VCDClient)

	vcdClient.lockParentVm(d)
	defer vcdClient.unLockParentVm(d)

	vm, vdc, err := getVm(vcdClient, d)
	if err != nil {
		return err
	}

	_, err = vm.GetInternalDiskById(d.Id(), false)
	if err != nil {
		return err
	}
	log.Printf("[TRACE] Internal Disk with id %s found", d.Id())

	var storageProfilePrt *types.Reference
	var overrideVmDefault bool

	storageProfileName := d.Get("storage_profile").(string)
	if storageProfileName != "" {
		storageProfile, err := vdc.FindStorageProfileReference(storageProfileName)
		if err != nil {
			return fmt.Errorf("[Error] error retrieving storage profile %s : %s", storageProfileName, err)
		}
		storageProfilePrt = &storageProfile
		overrideVmDefault = true
	} else {
		storageProfilePrt = vm.VM.StorageProfile
		overrideVmDefault = false
	}

	err = powerOffIfNeeded(d, vm)
	if err != nil {
		return err
	}

	diskSettingsToUpdate, err := vm.GetInternalDiskById(d.Id(), false)
	if err != nil {
		return err
	}

	log.Printf("[TRACE] Internal Disk with id %s found", d.Id())
	iops := int64(d.Get("iops").(int))
	isThinProvisioned := d.Get("thin_provisioned").(bool)
	diskSettingsToUpdate.Iops = &iops
	diskSettingsToUpdate.ThinProvisioned = &isThinProvisioned
	diskSettingsToUpdate.UnitNumber = d.Get("unit_number").(int)
	diskSettingsToUpdate.BusNumber = d.Get("bus_number").(int)
	diskSettingsToUpdate.AdapterType = internalDiskBusTypes[d.Get("bus_type").(string)]
	diskSettingsToUpdate.SizeMb = int64(d.Get("size_in_mb").(int))
	diskSettingsToUpdate.StorageProfile = storageProfilePrt
	diskSettingsToUpdate.OverrideVmDefault = overrideVmDefault

	_, err = vm.UpdateInternalDisks(vm.VM.VmSpecSection)
	if err != nil {
		return err
	}

	err = powerOnIfNeeded(d, vm)
	if err != nil {
		return err
	}

	log.Printf("[TRACE] Inernal Disk %s updated", d.Id())
	return nil
}
```

Reply:
Thank you for testing it and noticing on issue. We still have more issues to fix and improvements for this PR. Overall we glad that other things seems fine :)